doc/sphinx/src/nested_par_for.rst

```diff
@@ -86,3 +87,136 @@ GPUs.
 CPUs and so is the default on CPUs.
+
+
+Performance Considerations
+---------------------------
+
+Hierarchical parallelism can produce very performant code, but a
+deeper awareness of how hardware is mapped to threads is required to
+get optimal performance. Here we list a few strategies/considerations.
+
+* On CPU, with `SIMDFOR_INNER_LOOP` you may have trouble vectorizing
+  unless you help the compiler along. One way to do so is to work with
+  raw pointers to contiguous memory, rather than working with views
+  and strides. Even for stencil ops, if you can pull out pointers that
+  represent the different points on the stencil, this can help with
+  vectorization.
+* Similarly on CPUs, due to the cost of starting up a vector op,
+  vectorization will only be a performance win if there's enough work
+  in the inner loop. A minimum of 16 points is required for the op to
+  vectorize at all. Experience shows, however, that at least 64 is
+  really required to see big wins. One strategy for providing enough
+  vector cells in the inner loop is to do a 1D ``SIMDFOR`` inner loop
+  but combine the ``j`` and ``i`` indices by simply looping over the
+  contiguous memory in a rasterized plane on a block.
+* On GPUs, the outer loop typically maps to blocks, while the inner
+  maps to threads. To see good performance, you must both provide
+  enough work in the inner loop to create enough threads to fill in
+  CUDA terms a streaming multiprocessor (SM, equivalent to a Compute
+  Unit or CU on AMD GPUs) with multiple warps (or wavefronts for AMD)
+  to take advantage of pipelining and enough work in the outer loop to
+  create enough blocks to fill all SMs on the GPU divided by the
+  number of simultaneous streams. The number of warps in flight on the
+  inner loop per SM (which is related to "occupancy") will depend
+  positively on length of the inner loop and negatively on higher
+  shared memory usage (scratch pad memory in Kokkos parlance and Local
+  Data Store or LDS on AMD GPUs) and higher register usage. Note that
+  the number of SMs and the available shared memory and registers per
+  SM will vary between GPU architectures and especially between GPU
+  vendors.
+
+IndexSplit
+-------------
+
+To balance the CPU vs GPU hardware considerations of hierarchical
+parallelism, ``Parthenon`` provides a utility, the ``IndexSplit``
+class, defined in the ``utils/index_split.hpp`` header file and
+available in ``<parthenon/package.hpp>`` in the
+``parthenon::package::prelude`` namespace.
+
+In our experience ``IndexSplit`` is most beneficial when working with
+small meshblocks, especially in two dimensions. For small blocks, we
+want vectorized operations over contiguous memory for our innermost
+loop, but we want that loop to contain enough work for, e.g., vector
+ops to function. We have often found that the optimal split is to fuse
+j, and i into the inner loop and use k and blocks in the outer loop.
```

Comment:
@forrestglines here's the new paragraph

Reply:
I see! So does this give better performance on GPUs and on CPUs? The mention of vectorization suggests that this is a CPU improvement. In my mental model of GPUs launching fewer blocks vs. many blocks here could only have a detrimental performance if `k` is less than the number of SMs. How did this affect GPU performance in your experience?

Reply:
Could this functionality be inserted into `par_for` so that all local operations use this more optimal arrangement of teams?

Reply:
> I see! So does this give better performance on GPUs and on CPUs? The mention of vectorization suggests that this is a CPU improvement. 

Yeah this is a CPU optimization. I can add a comment saying so. I'm not sure what the optimal choice is for GPUs yet. We haven't explored it too much.

> Could this functionality be inserted into par_for so that all local operations use this more optimal arrangement of teams?

Possibly... I'd say we should put that in in a separate PR though. :)